### ISSUE TITLE
Segmentation error

### ISSUE THREAD

**Issue by Stormur** [cltk/lapos#1]
When trying to train a model, I get a segmentation error. This happens with the sample train.pos, too. When I try with the original files downloaded from http://www.logos.ic.i.u-tokyo.ac.jp/~tsuruoka/lapos/ I get no error.

Further, the latin_perseus_training.pos file is not provided here, as it should.

**Comment by anetschka** [comment]
Hi. Can you provide your command and the full error message?

**Comment by Stormur** [comment]
Hi!

From inside the clone git folder:

```
$ ./lapos-learn -m ./model <path to pos file>/latin_model_perseus.pos 
$ preparing for estimation...Errore di segmentazione (core dump creato)
```

Which means "Segmentation error (created core dump)".

As said, the very same command, with the same file and folders, works correctly with the "official" LaPos files.

**Comment by kylepjohnson** [comment]
Thanks @anetschka for the help.

@Stormur 
* To get our training set, use `curl -O https://raw.githubusercontent.com/cltk/latin_treebank_perseus/master/latin_training_set.pos`
* What OS are you using?
* How are you building the software? With `make`, like in the README?

**Comment by kylepjohnson** [comment]
Interestingly, the build server gets something similar. I don't remember this happening in the past: https://travis-ci.org/cltk/lapos/builds/458894035#L449

I'm training on my mac now and it's going fine.

**Comment by Stormur** [comment]
>     * To get our training set, use `curl -O https://raw.githubusercontent.com/cltk/latin_treebank_perseus/master/latin_training_set.pos`

OK, in the end I got it (`latin_model_perseus.pos` is just the same file renamed), it just didn't seem obvious to me. Maybe it should be mentioned in the Readme. 

>     * What OS are you using?

Ubuntu 18.04, recently upgraded from version 16.

>     * How are you building the software? With `make`, like in the README?

Yes, just a simple `make`, and everything (apparently) works.

**Comment by kylepjohnson** [comment]
Would you try installing the original source, available here?

If I recall correctly, the code in `master` ought to be exactly the same as the .tar.gz file they provide.

**Comment by Stormur** [comment]
Which source do you mean? I get the error when downloading and building the files from this github-project, from this branch. Is there another one?

**Comment by kylepjohnson** [comment]
From this page: http://www.logos.ic.i.u-tokyo.ac.jp/~tsuruoka/lapos/ (Get v. `0.1.2`)

Nobody within the CLTK wrote this, but we have re-hosted the code in a way that makes it a little more convenient for people to use.

**Comment by anetschka** [comment]
I, at least, googled this error: https://stackoverflow.com/questions/13654449/error-segmentation-fault-core-dumped
One of the suggestions is that the error is caused by RAM failure and this is actually plausible in a training scenario. This can be outruled by training on a more powerful machine or (if possible) by manually setting RAM usage for this process. One could also carefully debug the code of the training algorithm to see where exactly the error is thrown.

**Comment by kylepjohnson** [comment]
> One of the suggestions is that the error is caused by RAM failure and

I have wondered about that, but don't know C++ well enough to know. I agree, this might be the likely case (and also why it won't build on the low-RAM build server).

@Stormur Paste here the results of the command `free -mh`.

**Comment by Stormur** [comment]
> From this page: http://www.logos.ic.i.u-tokyo.ac.jp/~tsuruoka/lapos/ (Get v. `0.1.2`)
>
>This can be outruled by training on a more powerful machine or (if possible) by manually setting RAM usage for this process.

Right: as I stated before, using the original files from logos and training with them on the exact same data and laptop I had no problems and LaPos works perfectly afterwards, so it's not a problem of the machine. 

> @Stormur Paste here the results of the command `free -mh`.

```
                   total        used        free      shared  buff/cache   available
Mem:           7,7G        3,0G        1,7G        327M        3,0G        4,1G
Swap:           14G          0B         14G
```

**Comment by anetschka** [comment]
I am not a Linux pro, but it may be more insightful to monitor RAM usage during runtime, using the `top` or `htop` command from a second console window. This is not to say that your machine is broken ("problem of the machine"), rather, the hypothesis is that your training setting exceeds the RAM allocated to this process by the operating system. If you think the reason is something else, you can also debug the project code.

**Comment by kylepjohnson** [comment]
> as I stated before, using the original files 

My apologies, I forgot that.

I have pulled the original source and done a diff. Here is the result:

```
$ diff lapos lapos-0.1.2
Only in lapos: .git
Only in lapos: .gitignore
Only in lapos: .travis.yml
Only in lapos: README.md
diff lapos/crf.cpp lapos-0.1.2/crf.cpp
19c19
< const bool USE_BOS_EOS = false;
---
> const bool USE_BOS_EOS = false;  
29c29
< CRF_Model::CRF_Model()
---
> CRF_Model::CRF_Model() 
49,50c49,50
< 
< CRF_Model::~CRF_Model()
---
>  
> CRF_Model::~CRF_Model() 
417c417
<     //    for (vector<Path>::const_iterator i = vp.begin(); i != vp.end(); i++)
---
>     //    for (vector<Path>::const_iterator i = vp.begin(); i != vp.end(); i++) 
447c447
< 
---
>   
469c469
< 
---
>   
480c480
< #else
---
> #else    
506c506
< 
---
>   
651c651
< 
---
>   
655c655
< 
---
>   
664c664
< 
---
>   
687c687
<     // cerr << "warning: empty sentence" << endl;
---
>     cerr << "warning: empty sentence" << endl;
742c742
< 
---
>   
747c747
< 
---
>   
750c750
< 
---
>   
755c755
< 
---
>   
781c781
< 
---
>   
853c853
< 
---
>       
859c859
< 
---
>   
972c972
< 
---
>   
1007c1007
< void CRF_Model::decode_forward_backward(CRF_Sequence & s0,
---
> void CRF_Model::decode_forward_backward(CRF_Sequence & s0, 
1025c1025
< 
---
>   
1073c1073
< 
---
>   
1082c1082
< void
---
> void 
1141c1141
< 
---
>       
1180c1180
< 
---
>       
1222c1222
<   for (int i = 0; i < dim; i++) {
---
>   for (int i = 0; i < dim; i++) { 
1264c1264
< static void
---
> static void 
1271c1271
< 
---
>   
1298c1298
< 
---
>   
1353c1353
< 	k++;
---
> 	k++; 
1374a1375
> 
diff lapos/crf.h lapos-0.1.2/crf.h
42c42
<     features.push_back(f);
---
>     features.push_back(f);   
64c64
< 
---
>   
83,84c83,84
<   // enum { MAX_LABEL_TYPES = 50 };
<   const static int MAX_LABEL_TYPES = 2000;
---
>   enum { MAX_LABEL_TYPES = 50 };
>   //  const static int MAX_LABEL_TYPES = 1000;
91,93c91,93
< 
< private:
< 
---
>       
> private:  
>   
111c111
<     mefeature_type body() const {
---
>     mefeature_type body() const { 
141c141
< #else
---
> #else    
186c186
< 	//        v ^= *p << (i % 2);
---
> 	//        v ^= *p << (i % 2); 
201c201
< #else
---
> #else    
312c312
<   double lookahead_search(const Sequence & seq,
---
>   double lookahead_search(const Sequence & seq, 
315c315
< 			  const int max_depth,  const int depth,
---
> 			  const int max_depth,  const int depth, 
318c318
< 			  const bool follow_gold = false,
---
> 			  const bool follow_gold = false, 
321,323c321,323
< 		 const Sequence & seq,
< 		 const int start,
< 		 const std::vector<int> & history,
---
> 		 const Sequence & seq, 
> 		 const int start, 
> 		 const std::vector<int> & history, 
326,327c326,327
<   int update_weights_sub(const Sequence & seq,
< 			    std::vector<int> & history,
---
>   int update_weights_sub(const Sequence & seq, 
> 			    std::vector<int> & history, 
330,331c330,331
<   int update_weights_sub2(const Sequence & seq,
< 			    std::vector<int> & history,
---
>   int update_weights_sub2(const Sequence & seq, 
> 			    std::vector<int> & history, 
334,335c334,335
<   int update_weights_sub3(const Sequence & seq,
< 			    std::vector<int> & history,
---
>   int update_weights_sub3(const Sequence & seq, 
> 			    std::vector<int> & history, 
378c378
<         return p_edge_feature_id3[w * MAX_LABEL_TYPES * MAX_LABEL_TYPES * MAX_LABEL_TYPES + x * MAX_LABEL_TYPES * MAX_LABEL_TYPES + y * MAX_LABEL_TYPES + z]; }
---
>         return p_edge_feature_id3[w * MAX_LABEL_TYPES * MAX_LABEL_TYPES * MAX_LABEL_TYPES + x * MAX_LABEL_TYPES * MAX_LABEL_TYPES + y * MAX_LABEL_TYPES + z]; } 
384c384
<       return p_edge_feature_id2[x * MAX_LABEL_TYPES * MAX_LABEL_TYPES + y * MAX_LABEL_TYPES + z]; }
---
>       return p_edge_feature_id2[x * MAX_LABEL_TYPES * MAX_LABEL_TYPES + y * MAX_LABEL_TYPES + z]; } 
388c388
<       return p_edge_feature_id[l * MAX_LABEL_TYPES + r]; }
---
>       return p_edge_feature_id[l * MAX_LABEL_TYPES + r]; } 
Only in lapos: crf.o
Only in lapos: crfpos.o
Only in lapos: lapos
Only in lapos: lapos-eval
Only in lapos: lapos-learn
Only in lapos: lookahead.o
Only in lapos-0.1.2: model
Only in lapos-0.1.2: model_wsj00-18
Only in lapos-0.1.2: model_wsj02-21
Common subdirectories: lapos/samples and lapos-0.1.2/samples
Only in lapos: tokenize.o
```

@anetschka does this mean anything to you?

**Comment by anetschka** [comment]
@kylepjohnson No. I am not familiar with this project. If the reason for this error is not over-full RAM, it might be a real bug: C++, unlike more modern programming languages, requires programmers to work very carefully with memory. If this is quickly written scientific code, it might as well be very buggy (?). I located the error at least: It is in line 762 in file `crf.cpp`. So the method to debug is `make_feature_bag()`.

**Comment by anetschka** [comment]
I am still mainly guessing here, but one thing could be checked: The diff says

'const static int MAX_LABEL_TYPES = 2000;'
This is crf.h which I also cloned from this repo. Vocabulary size can make a huge difference in training performance, in particular if it is that large. @Stormur Do you need so many different pos labels? One quick thing could be to change that, recompile, and try again.

**Comment by kylepjohnson** [comment]
> located the error at least: It is in line 762 in file crf.cpp. So the method to debug is make_feature_bag().

OK. I think we are getting somewhere. In `crf.cpp` I see one meaningful diff (the others are, I think, just encoding changes):

```
---
>   
687c687
<     // cerr << "warning: empty sentence" << endl;
---
>     cerr << "warning: empty sentence" << endl;
742c742
< 
---
```

This looks like I only commented this warning out (my changes at the top).

**Comment by kylepjohnson** [comment]
... hit "enter" too soon ...

Another change is here, which I did purposefully, in order increase the max number of unique labels (this being necessary for a very detailed tagset, for such a highly inflected language).

```
diff lapos/crf.h lapos-0.1.2/crf.h
42c42
<     features.push_back(f);
---
>     features.push_back(f);   
64c64
< 
---
>   
83,84c83,84
<   // enum { MAX_LABEL_TYPES = 50 };
<   const static int MAX_LABEL_TYPES = 2000;
---
>   enum { MAX_LABEL_TYPES = 50 };
>   //  const static int MAX_LABEL_TYPES = 1000;
91,93c91,93
< 
< private:
< 
---
```

**Comment by kylepjohnson** [comment]
> @Stormur Do you need so many different pos labels? One quick thing could be to change that, recompile, and try again.

@anetschka  We were thinking the same thing. @Stormur  if the above above Latin POS set (and our greek version) do not need need to be as high as 2000, and this solves the problem, we can reduce it here.

**Comment by Stormur** [comment]
Now I am confused. Using the original source of LaPOS I noticed that the parameter for the maximum number of label had been changed to accommodate the big tagset of  Perseus; that is way I reduced the POS tags in my training set retaining only the first letter (here I have to apologise, I said that `latin_model_perseus.pos` was exactly the same as `latin_training_set.pos`), passing from 437 to only 14, under the threshold of 50 of the original LaPOS script (should these issues be mentioned in the README?).

Now I am fiddling with the lines you highlighted, but to no avail. Now I get the error:

`error: too many types of labels.`

I also found at the beginning of `crf.h` a commented
```
// use this option and change MAX_LABEL_TYPES below accordingly, if you want to use more than 255 labels
//#define USE_MANY_LABELS
```

**Comment by anetschka** [comment]
Hi Flavio (@Stormur),
it's very unfortunate that this produced a new error. It is thrown in crf.cpp, in CRF_Model:add_training_sample (apparently the method that collects the training data):

`if (s.label >= MAX_LABEL_TYPES - 2) {`
`    cerr << "error: too many types of labels." << endl;`
`    exit(1);`
`}`

It seems that you still have more labels in your data than allowed. What you could try is to set MAX_LABEL_TYPES to still another value. And the comment that you mentioned could also be helpful: 

`#ifdef USE_MANY_LABELS`
`typedef unsigned long mefeature_type;`

This indeed allows you to internally store more than 255 pos labels (by using a more powerful data type). How many different pos labels do you have?

Beyond that, it is really hard to say anything helpful without your complete command, input, and outputs. Monitoring RAM usage during run time, trying to train on a more powerful machine, and careful debugging (no fun with C++) are the options that I see.

Hope that helps,
a
